cmd/syncComparator.go

```diff
@@ -57,12 +65,30 @@ func (f *syncDestinationComparator) processIfNecessary(destinationObject StoredO
 	// if the destinationObject is present at source and stale, we transfer the up-to-date version from source
 	if present {
 		defer delete(f.sourceIndex.indexMap, destinationObject.relativePath)
-		if f.disableComparison || sourceObjectInMap.isMoreRecentThan(destinationObject) {
-			err := f.copyTransferScheduler(sourceObjectInMap)
-			if err != nil {
-				return err
+
+		if f.disableComparison {
+			return f.copyTransferScheduler(sourceObjectInMap)
+		}
+
+		if f.comparisonHashType != common.ESyncHashType.None() && sourceObjectInMap.entityType == common.EEntityType.File() {
+			switch f.comparisonHashType {
+			case common.ESyncHashType.MD5():
+				if !reflect.DeepEqual(sourceObjectInMap.md5, destinationObject.md5) {
+					azcopyScanningLogger.Log(pipeline.LogDebug, fmt.Sprintf("File %s overwritten due to differing hash.", destinationObject.relativePath))
+					// hash inequality = source "newer" in this model.
+					return f.copyTransferScheduler(sourceObjectInMap)
+				}
+			default:
+				panic("sanity check: unsupported hash type " + f.comparisonHashType.String())
```

Comment:
We should return error, instead of panic. We may specialize the error and quit all at once, but we should avoid panic in general.

Reply:
Panics are used exclusively for sanity checks, to make sure that values that _should_ for all reasons be correct at the given point in the code are still correct, and that garbage hasn't come in for some reason. They, for all intents and purposes, will never be hit. But if they are hit, we immediately know something is terribly wrong.